### ISSUE TITLE
[MCHECKSTYLE-384] Incompatibility to Checkstyle version >= 8.24  - Upgrade to 8.28

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-checkstyle-plugin#296]
**[Martin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkias)** opened **[MCHECKSTYLE-384](https://issues.apache.org/jira/browse/MCHECKSTYLE-384?redirect=false)** and commented

The latest `maven-checkstyle-plugin` is incompatible to Checkstyle version 8.24 and newer. The check for "LineLength" was moved from "TreeWalker" to "Checker".
For further details see "Breaking backward compatibility" under https://checkstyle.org/releasenotes.html#Release_8.24
 

```none
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (validate) on project top-secrect: Failed during checkstyle configuration: LineLength is not allowed as a child in Checker -> [Help 1]
[ERROR] ...
```

 

---

**Affects:** 3.1.0

**Remote Links:**
- [GitHub Pull Request #24
](https://github.com/apache/maven-checkstyle-plugin/pull/24)

0 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Benjamin Marwell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mampf86)** commented

If maven checkstyle plugin updated checkstyle, the user could either
* update his `checkstyle.xml` file
* use a lower dependency of puppycrawl checkstyle.
  Here's a how to: https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/upgrading-checkstyle.html

Updating checker to a higher version in this plugin would mean to release a new major version of this plugin. Sadly, the checkstyle maintainers did not do that, although they have a breaking change. That should have been checkstyle 9.x, not 8.24.

**Comment by jira-importer** [comment]
**[Roman Ivanov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=romani)** commented

Sorry for causing problems during updates.

Reason of why checkstyle is not following semantic versioning:
https://checkstyle.org/#Development_Cycle
https://github.com/checkstyle/checkstyle/issues/3709

8.X versions is full of heavy damages for compatibility, that we postponed for many years.

Plugin can do nothing with this change,  user config changes are unavoidable.
This issue should be closed.

> That should have been checkstyle 9.x

unfortunately no, minor version (it is minor update from checkstyle perspective, there are bunch of others updates happening that require config update) in checkstyle mean some non-bugs features are introduces. Other plugins(eclipse-cs , sonar) have to be updated even when new Check is introduced.
We will bump to 9.x probably when we update minimal jdk version or something really big.

**Comment by jira-importer** [comment]
**[Benjamin Marwell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mampf86)** commented

Actually, the Reporter got it wrong. He wrote:

>  The latest maven-checkstyle-plugin is incompatible to Checkstyle version 8.24

This is not entirely true, since this plugin is only a wrapper to some checkstyle version which happens to default to 8.19 atm.

That particular perspective seems also be shared on the maven dev mailing list. Although I would not want to break the plugin compatiblity with the default installations, this way of thinking eases the pain a little (given enough documentation). Other impacts are: https://issues.apache.org/jira/browse/MCHECKSTYLE-387 and possibly others.

**Comment by jira-importer** [comment]
**[Benjamin Marwell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mampf86)** commented

Just a quick idea which came to my mind:

Before the checkstyle dependency is updated, maybe rewrite this document:
https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/upgrading-checkstyle.html
 
Instead of writing "you CAN upgrade the checkstyle version" maybe even write "you SHOULD choose a checkstyle version explicitly"?

**Comment by jira-importer** [comment]
**[Martin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkias)** commented

Hi [bmarwell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bmarwell),

yes, I have update my pom-file and set checkstyle dependency version [explicit](https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/upgrading-checkstyle.html). Now it works. Thanks for the hint.

**Comment by jira-importer** [comment]
**[Jeremy Landis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hazendaz)** commented

The build is actually incompatible with IT tests as indicated here.  I'm working on fixing those now.  The change alone would indicate to those using this that there were issues post the points i'm marking per checkstyle change log.  That may help others.  I do just plan to target this Jira ticket.

**Comment by jira-importer** [comment]
**[Benjamin Marwell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mampf86)** commented

Ah now I get it: This project suffers from the same incompatibility problems as well. Sure. That didn't even came to my mind. I got confused between other users and this plugin as user itself. Maybe this should be pointed out more clearly.

**Comment by jira-importer** [comment]
**[Jeremy Landis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hazendaz)** commented

I'm going to open a new issue that is somewhat related.  Plugin 3.1.0 of checkstyle has a major regression in my opinion which started me down the path of cleaning things up.  The reports do not show the resulting aggregated rules that are violated.  If I drop back to 3.0.0, I get that up through checkstyle 8.24.  So to the outside user, maven checkstyle plugin 3.0.0 supports through checkstyle 8.24 with override without a problem.  It breaks after due to api changes.  Maven checkstyle plugin 3.1.0 regardless of the version it claims to use (8.19) through 8.28 fails to generate the aggregated rule violation section.  Its there, just empty.  I'm now reviewing 3.0.0 to 3.1.0 release to see if I can find where the problem is especially since version 3.0.0 actually works to a point.

**Comment by jira-importer** [comment]
**[Benjamin Marwell](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mampf86)** commented

Can you link the new issue please? And maybe add a integration test?

**Comment by jira-importer** [comment]
**[Enrico Olivelli](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=eolivelli)** commented

[Jeremy Landis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hazendaz) I can't set you as Assignee, btw I have merged your patch as e1e2f28620e2baed8351fea5fa524e98749df55c

I am going to upgrade to 8.29 and then cut a release
